File: DPM.m
%data generation
average_list=[15,19;-11.4,4;4.2,-17.4;-18.2,-6];
data=[;];
for i=1:size(average_list,1)
    x=normrnd(average_list(i,1),1,2000,1);
    y=normrnd(average_list(i,2),1,2000,1);
    new_data=cat(2,x,y);
    new_data(:,3)=i;%the 3rd column is the true k
    new_data(:,4)=0;%the 4th column is the prediction of k
    data=cat(1,data,new_data);
    
end
temp=randperm(size(data,1));
data=data(temp,:);% rearrange the points randomly



%DP
N=size(data,1);% # of points
%tag=zeros[N]; % initialization of every point's cluster
nk=[]; % # of points for each cluster
k_sum=[0,0]; % sum of each culster
mean=[0,0];
prob=[];

alpha=0.1;
D=2;% # of dimension of the data
for loop=1:10 % iteration for 100 times
    K=length(nk);
    for n=1:N
        
        point=data(n,[1,2]);
        k_old=data(n,4);
        if k_old>0
            nk(k_old)=nk(k_old)-1;
            k_sum(k_old,:)=k_sum(k_old,:)-data(n,[1,2]);
            
            if nk(k_old)==0 % remove the cluster which has no point
                nk(k_old)=[];
                k_sum(k_old,:)=[];
                mean(k_old,:)=[];
                for i=1:N
                    if data(i,4)>k_old
                        data(i,4)=data(i,4)-1;
                    end
                end
            elseif nk(k_old)<0
                'fuck'
            end
        end % remove the current point from its former cluster
        
        K=length(nk);
        prob=[];
        for k=1:K
            prob(k)=nk(k)/(N-1+alpha)/(2*pi)^(D/2)/exp(0.5*sum((point-mean(k,:)).^2));
        end
        prob(K+1)=alpha*sqrt(0.5)/(N-1+alpha)/(2*pi)^(D/2)/exp(0.25*sum(point.^2));
        K;
        prob;
        k_new=multisampling(prob);
        if k_new==(K+1)% assigned to a new cluster
            nk(k_new)=1;
            k_sum(k_new,:)=data(n,[1,2]);
        else %assigned to an existing cluster
            nk(k_new)=nk(k_new)+1;
            k_sum(k_new,:)=k_sum(k_new,:)+data(n,[1,2]);
        end
        data(n,4)=k_new;
        
        K=length(nk);
        for k=1:K
            k_mu=k_sum(k,:)./nk(k);
            mean(k,1)=normrnd(k_mu(1),1/nk(k),1,1);
            mean(k,2)=normrnd(k_mu(2),1/nk(k),1,1);
        end
    end
    K_count(loop)=K;
end



%evaluation
t=0;
for n=1:N
    if data(n,3)==data(n,4)
        t=t+1;
    end
end
'accuracy = '
t/N
mean
K
nk
%plot graph
figure
scatter(data(:,1),data(:,2),[],data(:,4),'.'),hold on,scatter(mean(:,1),mean(:,2),nk.*7,[1:K]', 'O'), hold off, title('RESULT')
    
        
            
        

File: DPM.py
import math
import random
import time


#####################    Data Generation    ################################
x=[]
y=[]
data=[]
average_list=[(2,2),(-2,2),(2,-2),(-2,-2)]
for item in average_list:
    x.append([random.gauss(item[0],1) for i in range(200)])
    y.append([random.gauss(item[1],1) for i in range(200)])

for k in range(len(average_list)):
    for i in range(200):
        data.append((x[k][i],y[k][i],k))
random.shuffle(data)

datafile=open('data','w')
for x,y,truek in data:
    print >>datafile,x,y,truek



    
def multisampling(Prob):
    Prob_sum=sum(Prob)
    Prob=[p/Prob_sum for p in Prob]
    pp=random.random()
    k=0
    P=0.0
    for p in Prob:
        P+=p
        if pp<=P:
            break
        else:
            k+=1
    return k

########################################    DPM    ############################################
time_begin = time.time()

N=len(data)
alpha=100
d=2
tag=[-1]*N
mean=[]
nk=[]
DK=[]
M_dist1=[]
M_dist2=[]
K_count=[]
for loop in range(10):
    prob=[]
    for n in range(N):
        x=data[n]
        k_old=tag[n]
        if k_old>=0:
            nk[k_old]-=1
        prob=[0]*(len(mean)+1)
        for k in range(len(mean)):
            #prob[k]=float(nk[k])/float(N-1+alpha)/(2*math.pi)**(d/2)/math.exp(0.5*sum([(x[i]-mean[k][i])**2 for i in range(d)]))
            if nk[k]==0:
                prob[k]=0.0
            else:
                prob[k]=math.log(float(nk[k])/float(N-1+alpha)/(2*math.pi)**(d/2))-(0.5*sum([(x[i]-mean[k][i])**2 for i in range(d)]))
                prob[k]=math.exp(prob[k])
        #prob[len(mean)]=alpha*math.sqrt(0.5)/(N-1+alpha)/(2*math.pi)**(d/2)/math.exp(0.25*sum([x[i]**2 for i in range(d)]))
        prob[len(mean)]=math.log(alpha*math.sqrt(0.5)/(N-1+alpha)/(2*math.pi)**(d/2))-(0.25*sum([x[i]**2 for i in range(d)]))
        prob[len(mean)]=math.exp(prob[len(mean)])
        k_new=multisampling(prob)
        if k_new==len(nk):
            nk.append(1)
        elif k_new>len(nk):
            print 'fuck'
        else:
            nk[k_new]+=1
        tag[n]=k_new
    for k in range(len(mean)+1):
        ck=0
        k_sum=[0.0,0.0]
        for n in range(N):
            if tag[n]==k:
                ck+=1
                k_sum[0]+=data[n][0]
                k_sum[1]+=data[n][1]
        k_mu=[k_sum[0]/(1+ck),k_sum[1]/(1+ck)]
        k_sigma=1.0/(1+ck)
        if k==len(mean) and ck!=0:
            mean.append([0.0,0.0])
        elif k==len(mean) and ck==0:
            break
        
        mean[k][0]=random.gauss(k_mu[0],k_sigma)
        mean[k][1]=random.gauss(k_mu[1],k_sigma)


    mean_new=[]
    nk_new=[]
    tag_new=[-1]*N
    k_new=0
    for k in set(tag):
        mean_new.append(mean[k])
        nk_new.append(nk[k])
        for n in range(N):
            if(tag[n]==k):
                tag_new[n]=k_new
        k_new+=1
    mean=mean_new
    nk=nk_new
    tag=tag_new

    K_count.append(len(nk))


    DK.append(len(nk)-math.log(float(alpha)/float((1+(N/alpha)))))
    M_dist1.append(sum([math.sqrt(sum([mean[k][i]**2 for i in range(d)])) for k in range(len(nk))]))
    M_dist2.append(sum([math.sqrt(sum([(data[n][i]-mean[tag[n]][i])**2 for i in range(d)])) for n in range(N)]))
    
time_last = time.time() - time_begin

print time_last

resultfile=open('result.m','w')
print >>resultfile, 'mean=['+';'.join('%s,%s' %(x,y) for x,y in mean)+']'
print >>resultfile, 'DK=',DK
print >>resultfile, 'M_dist1=',M_dist1
print >>resultfile, 'M_dist2=',M_dist2
print >>resultfile, 'tag=',tag
print >>resultfile, 'nk=',nk
print >>resultfile, 'K_count=',K_count


graphfile=open('graph.m','w')

print >>graphfile, '''result;'''
print >>graphfile, '''data = load('data');'''
print >>graphfile, 'figure'
print >>graphfile, '''scatter(data(:,1),data(:,2),[], tag, '.'), hold on, scatter(mean(:,1),mean(:,2),70*nk,[0:(length(nk)-1)]', 'O'), hold off, title('RESULT')'''
print >>graphfile, 'figure'
print >>graphfile, 'subplot(4,1,1)'
print >>graphfile, 'plot(K_count)'
print >>graphfile, '''title('K_count/iterations')'''
print >>graphfile, 'subplot(4,1,2)'
print >>graphfile, 'plot(DK)'
print >>graphfile, '''title('DK/iterations')'''
print >>graphfile, 'subplot(4,1,3)'
print >>graphfile, 'plot(M_dist1)'
print >>graphfile, '''title('M_dist1/iterations')'''
print >>graphfile, 'subplot(4,1,4)'
print >>graphfile, 'plot(M_dist2)'
print >>graphfile, '''title('M_dist2/iterations')'''



datafile.close()
resultfile.close()
graphfile.close()
##################################################################### evaluation###############################################################
t=0
f=0

for n in range(N):
    
    if data[n][d]==tag[n]:
        t+=1
    else:
        f+=1
print float(t)/float(N)
print 'done'

            
